pkg/crdmetrics/collector.go

```diff
@@ -0,0 +1,161 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crdmetrics
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/metric"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/open-telemetry/opentelemetry-operator/apis/v1alpha1"
+	"github.com/open-telemetry/opentelemetry-operator/apis/v1beta1"
+)
+
+type OpenTelemetryCollectorMetrics struct {
+	client       client.Client
+	observations []instancesView
+}
+
+func newOpenTelemetryCollectorMetrics(client client.Client) *OpenTelemetryCollectorMetrics {
+	return &OpenTelemetryCollectorMetrics{
+		client: client,
+	}
+}
+
+func (i *OpenTelemetryCollectorMetrics) Setup() error {
+	meter := otel.Meter(meterName)
+
+	obs, err := newObservation(meter,
+		receivers,
+		"Number of instances using certain receiver",
+		"type",
+		func(instance *collectorInfo) ([]string, bool) {
+			return instance.components.receivers, true
+		})
+	if err != nil {
+		return err
+	}
+	i.observations = append(i.observations, obs)
+
+	obs, err = newObservation(meter,
+		exporters,
+		"Number of instances using certain exporter",
+		"type",
+		func(instance *collectorInfo) ([]string, bool) {
+			return instance.components.exporters, true
+		})
+	if err != nil {
+		return err
+	}
+	i.observations = append(i.observations, obs)
+
+	obs, err = newObservation(meter,
+		processors,
+		"Number of instances using certain processor",
+		"type",
+		func(instance *collectorInfo) ([]string, bool) {
+			return instance.components.processors, true
+
+		})
+	if err != nil {
+		return err
+	}
+	i.observations = append(i.observations, obs)
+
+	obs, err = newObservation(meter,
+		extensions,
+		"Number of instances using certain extension",
+		"type",
+		func(instance *collectorInfo) ([]string, bool) {
+			return instance.components.extensions, true
+
+		})
+	if err != nil {
+		return err
+	}
+	i.observations = append(i.observations, obs)
+
+	obs, err = newObservation(meter,
+		mode,
+		"Collector deployment mode",
+		"mode",
+		func(instance *collectorInfo) ([]string, bool) {
+			return []string{instance.mode}, true
+
+		})
+	if err != nil {
+		return err
+	}
+	i.observations = append(i.observations, obs)
+	instruments := make([]metric.Observable, 0, len(i.observations))
+	for _, o := range i.observations {
+		instruments = append(instruments, o.Gauge)
+	}
+	_, err = meter.RegisterCallback(i.callback, instruments...)
+	return err
+}
+
+func (i *OpenTelemetryCollectorMetrics) callback(ctx context.Context, observer metric.Observer) error {
+	// Reset observations
+	for _, o := range i.observations {
+		o.reset()
+	}
+
+	v1alpha1collectors := &v1alpha1.OpenTelemetryCollectorList{}
```

Comment:
The thing with the reconciliation loop is that we don't have full control when the CR is deleted unless we use finalizers.

Reply:
true, however, we're not detecting deletes here either which would result in the same behavior. I think doing it like this feels like too much overheard. This also makes me think that we should use a counter which would allow us to not need to worry about this.

Reply:
May be I'm not getting something, because the reason we would need to detect a delete is because I was thinking on use a counter to increase/decrease depending on if we delete or create a new collector.

In this way we also not detecting a delete, but at least we are taking an snapshot of the cluster status at that point. I'm agree this could be a little bit more overheated but I'm not getting how we can do the same in the reconciliation loop.

Reply:
a counter only increases, so the way we would use it like this:
* each reconcile loop increase counter for collector/namespace combo for each of the metrics
* query in your backend based on a rate by doing max by collector/namespace combo

You no longer reconcile deleted collectors which means you no longer include count them. In your rate query they will stop appearing. I can explain in more detail if you'd like

Reply:
Ohh then, it will be an always incremental count and with the rate we can see when new instances are created? That make sense to me!

Reply:
precisely :)

Reply:
Are we planning to use collector/namespace combo as labels on the metric? This can lead to high cardinality metrics. @rubenvp8510 can telemeter ingest arbitrary labels for a given metric?

Reply:
The operator has as well a collector webhook which is called per create/update/delete.

Reply:
> Are we planning to use collector/namespace combo as labels on the metric? This can lead to high cardinality metrics. @rubenvp8510 can telemeter ingest arbitrary labels for a given metric?

I expressed this concern here: https://github.com/open-telemetry/opentelemetry-operator/issues/2829#issuecomment-2071417215 TBH I'm hessitating to collect collector_name/namespace because of that

Reply:
@pavolloffay i would be surprised if a user has more than 10 or even 100 distinct collector pools in a single cluster. If |~500| time series is a concern for some users they should simply define recording rules or metric relabel rules as part of their config. This isn't recording the metric for each pod which surely would lead to high cardinality.